docs/chymyst01.md

```diff
@@ -386,43 +395,164 @@ Sometimes it appears that repeating input molecules is the most natural way of e
+
+This kind of nondeterminism is the prime reason concurrency is widely regarded as a hard programming problem.
+
+`JoinRun` will actually refuse our attempted program and print an error message before running anything, immediately after we define the chemical laws:
+
+```scala
+val data = m[Int]
+val sum = m[Int]
+join (
+  run { case data(x) + sum(y) => sum(x+y) },
+  run { case sum(x) => println(s"sum = $x") }
+)
+```
+
+```
+Exception: In Join{data + sum => ...; sum => ...}: Unavoidable nondeterminism: reaction data + sum => ... is shadowed by sum => ...
+
+```
+
+The error message means that the reaction `sum => ...` will sometimes prevent `data + sum => ...` from running,
+and the programmer will have no control over this nondeterminism.
+
+The correct way of implementing this problem is to keep track of how many `data` molecules we already consumed,
+and to emit `done` when we reach the total expected number of the `data` molecules.
+The main change is that the `sum` molecule will now carry a tuple of two integers instead of a single integer.
+Here is an example solution with a single reaction:
+
+```scala
+val data = m[Int]
+val sum = m[(Int, Int)]
+join (
+  run { case data(x) + sum((y, remaining)) =>
+      val newSum = x + y
+      if (remaining == 1)  println(s"sum = $newSum")
+      else  sum((newSum, remaining-1)) 
+     }
+)
+data(5) + data(10) + data(150) // inject three `data` molecules
+sum((0, 3)) // expect "sum = 165" printed
+
+```
+
+There is no nondeterminism in these chemical laws, and no priority needs to be explicitly assigned.
+
+The chemical machine forces the programmer to design the chemistry in such a way that
+the order of running reactions is completely determined by the data on the available molecules.
+
 ## Summary so far
 
 The chemical machine requires for its description:
+
 - a list of defined molecules, together with their types;
 - a list of reactions involving these molecules as inputs, together with reaction bodies.
 
-The user can define reactions in one or more join definitions.
-One join definition encompasses all reactions that have some _input_ molecules in common.
+These definitions comprise the chemical laws of a concurrent program.
 
-After defining the molecules and specifying the reactions, the user can start injecting molecules into the soup.
+The user can define reactions in one or more join definitions.
+Each join definition encompasses all reactions that have some _input_ molecules in common.
+Different join definitions must have no input molecules in common.
 
 In this way, a complicated system of interacting concurrent processes can be specified through a particular set of chemical laws and reaction bodies.
 
+After defining the molecules and specifying the reactions, the user can start the program by injecting some initial molecules into the soup.
+
+The chemical laws fully specify which computations need to be performed for which data.
```

Comment:
"for which data" does not flow too well, maybe "for the corresponding data" or something of the kind.